agent.go

```diff
@@ -0,0 +1,461 @@
+	c.RLock()
+	defer c.RUnlock()
+
+	proc, exist := c.processes[pid]
+	if !exist {
+		return nil, fmt.Errorf("Process %s not found (container %s)", pid, c.id)
+	}
+
+	return proc, nil
+}
+
+func (p *pod) getContainer(id string) (*container, error) {
+	p.RLock()
+	defer p.RUnlock()
+
+	ctr, exist := p.containers[id]
+	if !exist {
+		return nil, fmt.Errorf("Container %s not found", id)
+	}
+
+	return ctr, nil
+}
+
+func (p *pod) setContainer(id string, ctr *container) {
+	p.Lock()
+	p.containers[id] = ctr
+	p.Unlock()
+}
+
+func (p *pod) deleteContainer(id string) {
+	p.Lock()
+	delete(p.containers, id)
+	p.Unlock()
+}
+
+func (p *pod) getRunningProcess(cid, pid string) (*process, *container, error) {
+	if p.running == false {
+		return nil, nil, fmt.Errorf("Pod not started")
+	}
+
+	ctr, err := p.getContainer(cid)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	status, err := ctr.container.Status()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if status != libcontainer.Running {
+		return nil, nil, fmt.Errorf("Container %s %s, should be %s", cid, status.String(), libcontainer.Running.String())
+	}
+
+	proc, err := ctr.getProcess(pid)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return proc, ctr, nil
+}
+
+func (p *pod) readStdio(cid, pid string, length int, stdout bool) ([]byte, error) {
+	proc, _, err := p.getRunningProcess(cid, pid)
+	if err != nil {
+		return nil, err
+	}
+
+	var file *os.File
+	if proc.termMaster != nil {
+		file = proc.termMaster
+	} else {
+		if stdout {
+			file = proc.stdout
+		} else {
+			file = proc.stderr
+		}
+	}
+
+	buf := make([]byte, length)
+
+	if _, err := file.Read(buf); err != nil {
+		return nil, err
+	}
+
+	return buf, nil
+}
+
+func (p *pod) initLogger() error {
+	agentLog.Logger.Formatter = &logrus.JSONFormatter{TimestampFormat: time.RFC3339Nano}
+
+	config := newConfig(defaultLogLevel)
+	if err := config.getConfig(kernelCmdlineFile); err != nil {
+		agentLog.WithError(err).Warn("Failed to get config from kernel cmdline")
+	}
+	config.applyConfig()
+
+	agentLog.WithField("version", Version).Info()
+
+	return nil
+}
+
+func (p *pod) initChannel() error {
+	// Check for vsock support.
+	vSockSupported, err := isAFVSockSupported()
+	if err != nil {
+		return err
+	}
+
+	if vSockSupported {
+		// Check if vsock socket exists. We want to cover the case
+		// where the guest OS can support vsock, but the runtime is
+		// still using virtio serial connection.
+		exist, err := vSockPathExist()
+		if err != nil {
+			return err
+		}
+
+		if exist {
+			// Fill the pod with vsock channel information.
+			p.channel = &vSockChannel{}
+
+			return nil
+		}
+	}
+
+	// Open serial channel.
+	file, err := openSerialChannel()
+	if err != nil {
+		return err
+	}
+
+	// Fill the pod with serial channel information.
+	p.channel = &serialChannel{
+		serialConn: file,
+	}
+
+	return nil
+}
+
+func (p *pod) startGRPC() (err error) {
+	var l net.Listener
+
+	if p.isVSock() {
+		l, err = vsock.Listen(1024)
```

Comment:
nit: constify the port number and make it part of agent API.

Reply:
Ok